### PR TITLE
Move retrystrategy facade to model config service

### PR DESCRIPTION
**Pull request by barrettj12**
This PR refactors the agent/retrystrategy facade to use a model config service instead of state.
- The `st`, `model` and `resources` fields of `RetryStrategyAPI` were removed and replaced with `modelConfigService` and `watcherRegistry` as appropriate.
- Calls to `model.ModelConfig` were replaced by calls to `modelConfigService.ModelConfig`.
- Calls to `model.WatchForModelConfigChanges` were replaced by calls to `modelConfigService.Watch`, with the logic changed appropriately to handle the new watcher type.
- A `service.go` file was added defining the `interface ModelConfigService`.
- We generated a mock for `ModelConfigService` using gomock. The test suite was changed so that we no longer use `ApiServerSuite`, instead we use the generated mock.
- A new `NewRetryStrategyAPI` method was create to help enable testing.
- The `export_test.go` file was deleted as it became unnecessary.
- The `pacakge_test.go` was renamed to `package_test.go`.
- Drive-by: remove a code block in the uniter that was killing the leadership tracker worker and causing container crashes. This was only relevant for podspec charms and should be removed now. See https://github.com/juju/juju/pull/17484

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Deploy an application, change the value of the `automatically-retry-hooks` model config key, and watch debug-log to see the retrystrategy / uniter workers restart with the new config.

```bash
juju bootstrap minikube c
juju add-model m
# set logging config on the model
juju model-config logging-config="<root>=WARNING;juju.worker.retrystrategy=DEBUG;juju.worker.uniter=DEBUG"
# deploy an application, any application
juju deploy tiny-bash
```

Toggle the value of `automatically-retry-hooks`:
```console
$ juju model-config automatically-retry-hooks
True
$ juju model-config automatically-retry-hooks=False
```

Watch debug-log on the model to see the uniter restart. This should be instant.
```
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.retrystrategy bouncing retrystrategy worker to get new values
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener stopping
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter unit "tiny-bash/0" shutting down: catacomb 0xc0004ad408 is dying
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener stopped
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter starting local juju-exec listener on {unix /var/lib/juju/agents/unit-tiny-bash-0/run.socket <nil>}
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter unit "tiny-bash/0" started
unit-tiny-bash-0: 11:46:31 DEBUG juju.worker.uniter juju-exec listener running
unit-tiny-bash-0: 11:46:31 INFO juju.worker.uniter hooks are retried true
...
```

## Links

**Jira card:** JUJU-6098

### CODE REVIEW

**Comment by barrettj12** [comment]
We're seeing strange behaviour on k8s that changing the `automatically-retry-hooks` key is triggering container restarts / crashes

**Comment by barrettj12** [comment]
QA'd and seems to work. The modeloperator container is crashing on k8s, but that seems to happen when you create the model, so it's not related to this change.
<details>

```
panic: InsecureKeyProfile cannot be used outside tests

goroutine 23 [running]:
github.com/juju/juju/internal/pki/test.InsecureKeyProfile()
        /home/jordan/git/juju/juju/retrystrategy-modelservice/internal/pki/test/keyprofile.go:20 +0x7d
github.com/juju/juju/internal/worker/apiservercertwatcher.newPKIAuthorityWorker()
        /home/jordan/git/juju/juju/retrystrategy-modelservice/internal/worker/apiservercertwatcher/authorityworker.go:28 +0x1e
github.com/juju/juju/internal/worker/apiservercertwatcher.NewAuthorityWorker({0xc00067cd80?, 0x5066364?})
        /home/jordan/git/juju/juju/retrystrategy-modelservice/internal/worker/apiservercertwatcher/authorityworker.go:24 +0xf
github.com/juju/juju/cmd/jujud/agent/modeloperator.Manifolds.Manifold.func3({0x50a21cc?, 0x50d5c35?}, {0x5849fa0, 0xc00067cd80})
        /home/jordan/git/juju/juju/retrystrategy-modelservice/internal/worker/apiservercertwatcher/manifold.go:43 +0xa4
github.com/juju/worker/v4/dependency.(*Engine).runWorker.func1()
        /home/jordan/go/pkg/mod/github.com/juju/worker/v4@v4.0.0/dependency/engine.go:530 +0x376
github.com/juju/worker/v4/dependency.(*Engine).runWorker.func2()
        /home/jordan/go/pkg/mod/github.com/juju/worker/v4@v4.0.0/dependency/engine.go:534 +0x4a
github.com/juju/worker/v4/dependency.(*Engine).runWorker(0xc0006cb9e0, {0x50a21cc, 0x13}, 0x0?, 0x0?, 0xc00067cd80)
        /home/jordan/go/pkg/mod/github.com/juju/worker/v4@v4.0.0/dependency/engine.go:571 +0xdf
created by github.com/juju/worker/v4/dependency.(*Engine).requestStart in goroutine 21
        /home/jordan/go/pkg/mod/github.com/juju/worker/v4@v4.0.0/dependency/engine.go:432 +0x81b
```
</details>

File: internal/worker/uniter/uniter.go [hunk 1]
```diff
@@ -277,11 +277,6 @@ func newUniter(uniterParams *UniterParams) func() (worker.Worker, error) {
 				return u.loop(uniterParams.UnitTag)
 			},
 		}
-		if u.modelType == model.CAAS {
```
Comment (hpidcock) [inline internal/worker/uniter/uniter.go:280]: For those interested: https://github.com/juju/juju/pull/17484

File: go.sum [hunk 2]
```diff
@@ -152,8 +152,8 @@ github.com/canonical/go-flags v0.0.0-20230403090104-105d09a091b8 h1:zGaJEJI9qPVy
 github.com/canonical/go-flags v0.0.0-20230403090104-105d09a091b8/go.mod h1:ZZFeR9K9iGgpwOaLYF9PdT44/+lfSJ9sQz3B+SsGsYU=
 github.com/canonical/lxd v0.0.0-20231214113525-e676fc63c50a h1:Tfo/MzXK5GeG7gzSHqxGeY/669Mhh5ea43dn1mRDnk8=
 github.com/canonical/lxd v0.0.0-20231214113525-e676fc63c50a/go.mod h1:UxfHGKFoRjgu1NUA9EFiR++dKvyAiT0h9HT0ffMlzjc=
-github.com/canonical/pebble v1.11.0 h1:Xl4JAK0Sc3197jUZJfRiUapUpTW8+Lh42AwWBS+PaeY=
-github.com/canonical/pebble v1.11.0/go.mod h1:BXpt85cFqrBgACeVRrTQ7JxZIdnGILv32V7mAfDcGFc=
+github.com/canonical/pebble v1.11.1-0.20240606212938-6688cb766465 h1:xp7Hfvxdcy3xOfICnPgLI1xX+YxcBquve67wCn9Bnw4=
```
Comment (hpidcock) [inline go.sum:155]: This will be fixed once the pebble release is out and is OK to land in main.

**Review by hpidcock — APPROVED** [review]
LGTM. I'm glad we found that bug in the uniter.

**Pull request merged**